src/box/memtx_tree.cc

```diff
@@ -886,7 +886,17 @@ memtx_tree_index_random(struct index *base, uint32_t rnd, struct tuple **result)
 	struct memtx_tree_index<USE_HINT> *index =
 		(struct memtx_tree_index<USE_HINT> *)base;
 	struct memtx_tree_data<USE_HINT> *res = memtx_tree_random(&index->tree, rnd);
-	*result = res != NULL ? res->tuple : NULL;
+	struct txn *txn = in_txn();
+	struct space *space = space_by_id(base->def->space_id);
+	if (res == NULL) {
+		*result = NULL;
+		memtx_tx_track_gap(txn, space, base, NULL, ITER_GE, NULL, 0);
+		return 0;
+	}
+	bool is_multikey = base->def->key_def->is_multikey;
+	uint32_t mk_index = is_multikey ? (uint32_t)res->hint : 0;
+	*result =
+		memtx_tx_tuple_clarify(txn, space, res->tuple, base, mk_index);
```

Comment:
Invisible tuple?

Reply:
Discussed with @alyapunov: let's first check if the space is empty from the transaction's perspective, and if it's not repeatedly call `random` and clarify result, until we get a non-empty one.

We do not care about performance here, since all operations in context of
transaction management currently have O(number of dirty tuples)
complexity.